sources/corepackUtils.ts

```diff
@@ -104,15 +106,46 @@ export async function findInstalledVersion(installTarget: string, descriptor: De
   return bestMatch;
 }
 
+export function isNotURLDescriptor(descriptor: Descriptor): descriptor is SupportedPackageManagerDescriptor {
+  return !URL.canParse(descriptor.range);
+}
+
+export function isSupportedPackageManagerLocator(locator: Locator): locator is SupportedPackageManagerLocator {
+  return !locator.isURL;
```

Comment:
do we need this `isURL` property? Why not just do a `URL.canParse(locator.reference)` like we do with descriptors?

Reply:
It's to avoid calling `URL.canParse` more than one – it's a way to cache the value rather than re-compute it every time we need it.

Reply:
That feels a little unnecessary imo (and tbh, I'd probably just check whether the string `startsWith('https://')` rather than validate the full correctness - feels a little too susceptible to typos).

Reply:
> what if I add a character that makes the string an invalid url? should it go in the semver path?

In all likeliness, an invalid URL due to a typo won't be a valid semver version either, so I don't think it matters.



> validating everything feels a little too susceptible to typos

Well we can only work with well formed reference (either semver or URL), so I'm not sure where you going with that, I don't see how we could be forgiving, we need full correctness.



> That feels a little unnecessary imo, we're only going to make a single such check per command, I doubt it'd have a significant perf impact compared to the complexity cost.

I feel like I'm missing something, I don't see how it can be simplified – because otherwise TS cannot differenciate between the URL-based spec and the semver-based ones.

Reply:
> because otherwise TS cannot differenciate between the URL-based spec and the semver-based ones.

Why does it need to differenciate? That's the same thing we do in Yarn: we treat all ranges the same (as string), and we just branch their behaviour by runtime pattern matching. I don't see what we gain by using strict typing here.